src/chapter_3/PDAs.md

```diff
@@ -0,0 +1,315 @@
+
+#[program]
+mod puppet_master {
+    use super::*;
+    pub fn pull_strings(ctx: Context<PullStrings>, bump: u8, data: u64) -> ProgramResult {
+        let cpi_program = ctx.accounts.puppet_program.to_account_info();
+        let cpi_accounts = SetData {
+            puppet: ctx.accounts.puppet.to_account_info(),
+            authority: ctx.accounts.authority.to_account_info()
+        };
+        let bumps = &[bump][..];
+        let signer = &[&[bumps][..]];
+        let cpi_ctx = CpiContext::new_with_signer(cpi_program, cpi_accounts, signer);
+        puppet::cpi::set_data(cpi_ctx, data)
+    }
+}
+
+#[derive(Accounts)]
+pub struct PullStrings<'info> {
+    #[account(mut)]
+    pub puppet: Account<'info, Data>,
+    pub puppet_program: Program<'info, Puppet>,
+    pub authority: UncheckedAccount<'info>
+}
+```
+
+The `authority` account is now an `UncheckedAccount` instead of a `Signer`. When the puppet-master is invoked, the `authority` pda is not a signer yet so we mustn't add a check for it. We just care about the puppet-master being able to sign so we don't add any additional seeds. Just a bump that is calculated off-chain and then passed to the function. This is the new `puppet.ts`:
+```ts
+import * as anchor from '@project-serum/anchor';
+import { web3 } from '@project-serum/anchor/';
+import { Program } from '@project-serum/anchor';
+import { Puppet } from '../target/types/puppet';
+import { PuppetMaster } from '../target/types/puppet_master';
+import { expect } from 'chai';
+
+describe('puppet', () => {
+  anchor.setProvider(anchor.Provider.env());
+
+  const puppetProgram = anchor.workspace.Puppet as Program<Puppet>;
+  const puppetMasterProgram = anchor.workspace.PuppetMaster as Program<PuppetMaster>;
+
+  const puppetKeypair = web3.Keypair.generate();
+
+  it('Does CPI!', async () => {
+    const [puppetMasterPDA, puppetMasterBump] = await web3.PublicKey
+      .findProgramAddress([], puppetMasterProgram.programId);
+
+    await puppetProgram.rpc.initialize(puppetMasterPDA, {
+      accounts: {
+        puppet: puppetKeypair.publicKey,
+        user: anchor.getProvider().wallet.publicKey,
+        systemProgram: web3.SystemProgram.programId,
+      },
+      signers: [puppetKeypair]
+    });
+
+    await puppetMasterProgram.rpc.pullStrings(puppetMasterBump, new anchor.BN(42),{
+      accounts: {
+        puppetProgram: puppetProgram.programId,
+        puppet: puppetKeypair.publicKey,
+        authority: puppetMasterPDA
+      ,
+    }});
+
+    expect((await puppetProgram.account.data
+      .fetch(puppetKeypair.publicKey)).data.toNumber()).to.equal(42);
+  });
+});
+```
+
+The `authority` is no longer a randomly generated keypair but a PDA derived from the puppet-master program. This means the puppet-master can sign with it which it does inside `pullStrings`. It's worth noting that our implementation also allows non-canonical bumps but again because we are only interesting in being able to sign we don't care which bump is used.
```

Comment:
Would be nice to add a comment saying that the two uses of pdas can be used together, i.e., for both signing and determinstic data.

Reply:
just to be clear, you mean that a single PDA can be used in an instruction as both a state account and a signer account for a CPI?

Reply:
correct